File: android/app/src/main/java/com/myvu/client/app/feature/HudMessageQueue.java
package com.myvu.client.app.feature;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

/**
 * Splits and queues long text responses into readable sentence-based HUD frames.
 */
public class HudMessageQueue {
    private static final int MAX_HUD_FRAME_LEN = 80;
    private final Queue<String> pendingFrames = new ArrayDeque<>();

    public void enqueue(String fullText) {
        pendingFrames.clear();
        if (fullText == null || fullText.trim().isEmpty()) return;

        List<String> sentences = splitIntoSentences(fullText.trim());
        for (String sentence : sentences) {
            if (sentence.length() <= MAX_HUD_FRAME_LEN) {
                pendingFrames.add(sentence);
            } else {
                // Chunk long sentences by word boundaries
                String[] words = sentence.split("\\s+");
                StringBuilder sb = new StringBuilder();
                for (String word : words) {
                    if (sb.length() + word.length() + 1 > MAX_HUD_FRAME_LEN) {
                        pendingFrames.add(sb.toString().trim());
                        sb.setLength(0);
                    }
                    if (sb.length() > 0) sb.append(" ");
                    sb.append(word);
                }
                if (sb.length() > 0) {
                    pendingFrames.add(sb.toString().trim());
                }
            }
        }
    }

    public boolean hasNext() {
        return !pendingFrames.isEmpty();
    }

    public String pollNext() {
        return pendingFrames.poll();
    }

    public void clear() {
        pendingFrames.clear();
    }

    private static List<String> splitIntoSentences(String text) {
        List<String> result = new ArrayList<>();
        String[] parts = text.split("(?<=[\\.\\?\\!\\n])\\s+");
        for (String part : parts) {
            String trimmed = part.trim();
            if (!trimmed.isEmpty()) {
                result.add(trimmed);
            }
        }
        return result;
    }
}

File: android-kotlin/app/src/main/java/com/myvu/client/app/feature/HudMessageQueue.kt
package com.myvu.client.app.feature

import java.util.ArrayDeque
import java.util.Queue

/**
 * Splits and queues long text responses into readable sentence-based HUD frames.
 */
class HudMessageQueue {
    companion object {
        private const val MAX_HUD_FRAME_LEN = 80

        private fun splitIntoSentences(text: String): List<String> {
            val result = ArrayList<String>()
            val parts = text.split(Regex("(?<=[\\.\\?\\!\\n])\\s+"))
            for (part in parts) {
                val trimmed = part.trim()
                if (trimmed.isNotEmpty()) {
                    result.add(trimmed)
                }
            }
            return result
        }
    }

    private val pendingFrames: Queue<String> = ArrayDeque()

    fun enqueue(fullText: String?) {
        pendingFrames.clear()
        if (fullText.isNullOrBlank()) return

        val sentences = splitIntoSentences(fullText.trim())
        for (sentence in sentences) {
            if (sentence.length <= MAX_HUD_FRAME_LEN) {
                pendingFrames.add(sentence)
            } else {
                val words = sentence.split(Regex("\\s+"))
                val sb = StringBuilder()
                for (word in words) {
                    if (sb.length + word.length + 1 > MAX_HUD_FRAME_LEN) {
                        pendingFrames.add(sb.toString().trim())
                        sb.setLength(0)
                    }
                    if (sb.isNotEmpty()) sb.append(" ")
                    sb.append(word)
                }
                if (sb.isNotEmpty()) {
                    pendingFrames.add(sb.toString().trim())
                }
            }
        }
    }

    fun hasNext(): Boolean = !pendingFrames.isEmpty()

    fun pollNext(): String? = pendingFrames.poll()

    fun clear() {
        pendingFrames.clear()
    }
}
